tutorials/untyped/untyped.md

```diff
@@ -89,39 +103,55 @@ TL;DR: Untyped objects should be allocated in order of size, largest first, to a
 parameter is the untyped capability being invoked to do the retype.
 
 #### Types
 
 Each object in seL4 has a specific type, and all of the type constants can be found in `libsel4`. Some types
 are architecture specific, while others are general across architectures. In the example above, we create a
-new untyped object, which can be further retyped. The type of the created object determines the invocations
-that can be made on that object. For example, were we to retype the object into a thread control block
-(TCB -- `seL4_TCBObject`), we could then perform TCB invocations on the object, including `seL4_TCB_SetPriority`.
+new untyped capability, which can be further retyped. All other object type parameters create kernel objects *and*
+capabilities to them, and the type of the created object (and capability) determines the invocations
+that can be made on that capability. For example, were we to retype into a thread control block
+(TCB -- `seL4_TCBObject`), we could then perform TCB invocations on the capability to the new TCB object, including `seL4_TCB_SetPriority`.
 
 #### Size
 
-The size argument determines the size of the new object and is specified in `size_bits`, where the corresponding object
- size is the exponentiation, where `size_bits` is the exponent and the base is 2. In other words,
- the object size is 2<sup>`size_bits`</sup> in bytes.
-
-All seL4 objects must be powers of 2 in size, however some objects are fixed-size while other objects are
- variable in size. The size parameter is *ignored* for fixed-sized objects and only respected by the kernel for
- variably-sized objects. Variably sized objects include untypeds (`seL4_UntypedObject`), CNodes (`seL4_CapTableObject`)
- and frame objects, of which the latter are architecture dependent.
+The size argument determines the size of the new object and is specified in
+`size_bits`. In general in seL4, if sizes are measured in bits, they are powers
+of two. "Bits"  does not refer to the number of bits the object takes up, but
+the bit-width that is needed to describe its contents. An object of bit size `n`
+measures 2<sup>`n`</sup> bytes. Also, generally in seL4, objects and memory
+regions are aligned to their size, i.e. an `n`-bit object is aligned to
+2<sup>`n`</sup> bytes, or, equivalently, the address of the object has 0 as its
+`n` bottom bits.
+
+For `retype`, it is enough to remember that the parameter `size_bits` means the
+object will measure 2<sup>`size_bits`</sup> bytes.
+
+Most objects in seL4 are of a fixed size, but some objects are variable in size. The
+kernel knows the size of fixed-size objects and the retype operation will
+*ignore* the size parameter for those objects. It will only use the parameter
+for variably-sized objects. Variably-sized objects include CNodes
+(`seL4_CapTableObject`) and frame objects, of which the latter are architecture
+dependent. Viewing untyped memory as an object (`seL4_UntypedObject`), the size
```

Comment:
Are there any variable-sized frame objects anymore? I did a quick look and couldn't find any. It looks to me like there are distinct object types for each page size.

Reply:
Looking at the implementation of `getObjectSize`, I think the only other object that is variable sized is the SchedContext object.

Reply:
Yes, frames definitely aren't variable, I just reshuffled those bits without thinking about them. Will fix.